src/sdkUtils/bridgeFromSolana.ts

```diff
@@ -0,0 +1,568 @@
+import type { Transaction } from "@solana/web3.js"
+import { encodeFunctionData, hexToBytes, toHex } from "viem"
+import { SDK_NAME } from "../bitcoinUtils/constants"
+import {
+  getEVMTokenContractInfo
+} from "../evmUtils/contractHelpers"
+import { sendMessageAbi } from "../evmUtils/contractMessageHelpers"
+import { metaTokenToCorrespondingStacksToken } from "../metaUtils/peggingHelpers"
+import { AnchorWrapper } from "../solanaUtils/anchorWrapper"
+import { getSolanaConfigs, getSolanaSupportedRoutes } from "../solanaUtils/getSolanaSupportedRoutes"
+import { isSupportedSolanaRoute } from "../solanaUtils/peggingHelpers"
+import { getStacksTokenContractInfo } from "../stacksUtils/contractHelpers"
+import {
+  checkRouteValid,
+  KnownRoute_FromSolana_ToBitcoin,
+  KnownRoute_FromSolana_ToBRC20,
+  KnownRoute_FromSolana_ToEVM,
+  KnownRoute_FromSolana_ToRunes,
+  KnownRoute_FromSolana_ToSolana,
+  KnownRoute_FromSolana_ToStacks,
+  KnownRoute_FromSolana_ToTron,
+} from "../utils/buildSupportedRoutes"
+import {
+  InvalidMethodParametersError,
+  UnsupportedBridgeRouteError,
+} from "../utils/errors"
+import { assertExclude, checkNever } from "../utils/typeHelpers"
+import {
+  _knownChainIdToErrorMessagePart,
+  KnownChainId,
+  KnownTokenId,
+} from "../utils/types/knownIds"
+import {
+  ChainId,
+  evmNativeCurrencyAddress,
+  SDKNumber,
+  TokenId,
+  type EVMAddress
+} from "./types"
+import { SDKGlobalContext } from "./types.internal"
+import { contractAssignedChainIdFromKnownChain } from "../lowlevelUnstableInfos"
+
+export type BridgeFromSolanaInput = {
+  fromChain: ChainId
+  toChain: ChainId
+  fromToken: TokenId
+  toToken: TokenId
+  fromAddress: string
+  toAddress: string
+  /**
+   * **Required** when `toChain` is one of bitcoin chains
+   */
+  toAddressScriptPubKey?: Uint8Array
+  /**
+   * The token account that holds the tokens to be bridged
+   */
+  senderTokenAccount: string
```

Comment:
`@solana/spl-token` has a function `getAssociatedTokenAddress` to calculate the token account from `mintAddress` and `ownerAddress`